update StruveL and StruveM functions

heltonmc opened the pull request:
This PR seeks to update several of the implementations. Mainly it adds much faster and accurate version of the power series and large argument expansion for all the functions.

In the case of `struveL` and `struveM` the power series and large argument expansion are enough to compute for every value. This is actually more accurate than using the integral representation and 10x faster and doesn't allocate.
```julia
julia> @benchmark Struve.L(1.5, x) setup=(x=rand()*100)
BenchmarkTools.Trial: 7988 samples with 195 evaluations.
 Range (min … max):  477.528 ns … 14.556 μs  ┊ GC (min … max): 0.00% … 70.35%
 Time  (median):       3.113 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):     3.202 μs ±  1.311 μs  ┊ GC (mean ± σ):  2.48% ±  6.60%

  ▄    ▄▂              ▂▄▃   ▅       ▁▄▁ ▇▃          ▅█▁       ▂
  █▁▁▁▁███▁▁▁▁▁▁▁▁▁▁▁▁▁███▃▅▅█▇▃▁▃▆▁▁███▄██▇▅▁▃▇▆▁▁▄▁███▆▅▃▃▃█ █
  478 ns        Histogram: log(frequency) by time      4.44 μs <

 Memory estimate: 192 bytes, allocs estimate: 7.

julia> @benchmark Struve.struveL(1.5, x) setup=(x=rand()*100)
BenchmarkTools.Trial: 10000 samples with 918 evaluations.
 Range (min … max):  113.034 ns …  1.135 μs  ┊ GC (min … max): 0.00% … 78.96%
 Time  (median):     236.073 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   210.256 ns ± 38.950 ns  ┊ GC (mean ± σ):  0.04% ±  0.79%

                                                         █▃▄▃   
  ▂▂▂▂▃▂▂▂▂▂▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▃▂▂▂▂▂▂▂▂▂▁▁▁████▄ ▃
  113 ns          Histogram: frequency by time          245 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

It also gives much more accurate results than using QuadGk though we could increase the tolerances there.
```julia
ulia> Struve.struveL(1.0, 10.0)
2670.358285208483

julia> Struve.L(1.0, 10.0)
2670.3582852082404

julia> Struve.struveL(big"1.0", big"10.0")
2670.358285208482969378435752177230712181823683454750407620682833591038763364245
```

The case for `StruveH` and `StruveK` are a little more difficult though we can still use the large order expansion and power series however the power series can not be as indiscriminately used as it suffers from cancellation in some regions. So we will need a third algorithm to compute that range which I'll address in a different PR.

Let me add some tests.

heltonmc committed:
update expansions

heltonmc committed:
add struveK and struveM

heltonmc committed:
add doc

heltonmc committed:
update struveL and struveM

heltonmc committed:
export functions

heltonmc commented:
I think maybe for now we can use the QuadGK form in that third region for StruveH and StruveK until we drop that algorithm completely with a better method. I can make that edit so this isn't so piecewise for all the different algorithms.

Of course for complex results the large argument expansion can't be applied so might just have to define the integral representation instead for now.

heltonmc committed:
add tests for struveL

heltonmc committed:
default to integral

heltonmc committed:
fix premature type demotions

heltonmc committed:
add tests for struve M

heltonmc committed:
add struveh and struvek tests

heltonmc committed:
fix integral form

heltonmc commented:
Ok I've modified this a little bit to fall back to using the adaptive numerical integral if the power series or large argument expansion don't provide good results. Though we need to be careful when we employ any identities to get the other forms because that subtraction is prone to cancellation even when using the adaptive integral forms.

I think it might be best to then piece together other algorithms as we get those separately to eventually remove the quadgk dependency completely.

I think there's a few other design decisions as well. Is there a reason you chose not to export these functions? Internally I think it might make sense to have it as `struvek` or `struveK` so it can be distinguished from `besselk`.

heltonmc committed:
add CI

heltonmc committed:
fix CI

heltonmc committed:
handle subnormals

heltonmc committed:
Delete .travis.yml

heltonmc commented:
Ok I've also added CI on this to help test. As of now all tests are passing. I've added the tests mentioned in https://github.com/gwater/Struve.jl/issues/2 though I believe other tests are better handled using higher precision computation of the power series.

gwater commented on src/integral.jl:
```diff
@@ -2,31 +2,41 @@
+_M0_integral(z) = -(2 / T(π)) * integrate(ϑ -> exp(-z * cos(ϑ)), zero(T), T(π)/2)
+
+
+function _M_integral(ν, z)
```
I think `T` is missing from the method signature here

heltonmc replied:
Yes - great catch!

gwater commented on src/integral.jl:
```diff
@@ -2,31 +2,41 @@
-_M0_integral(z) = -(2 / pi) * integrate(ϑ -> exp(-z * cos(ϑ)), 0, 0.5pi)
-_M_integral(ν, z) = -2(0.5z)^ν / (sqrt(pi) * gamma(ν + 0.5)) *
-    integrate(t -> exp(-z * t) * (1 - t^2)^(ν - 0.5), 0, 1)
+_M0_integral(z) = -(2 / T(π)) * integrate(ϑ -> exp(-z * cos(ϑ)), zero(T), T(π)/2)
```
I think `T` is missing from the method signature here

heltonmc replied:
Should be fixed now

gwater commented on src/struveH.jl:
```diff
@@ -0,0 +1,73 @@
+struveH(v, x) = _H_integral(v, x)
+
+_struveH(v, x::AbstractFloat) = _H_integral(v, x)
+_struveH(v::Float32, x::Float32) = Float32(_struveH(Float64(v), Float64(x)))
```
I feel like this sort of convenience implementation kind of misleads users about the capabilities of the package. It's probably better to just throw an MethodError and admit there isnt a 32-bit / 16-bit implementation yet

heltonmc replied:
I'm happy with whatever you think is best. I have just updated `struveH` and `struveL` to have explicit `Float32` routines. These routines are not prone to cancellation in the sums so they can be used pretty indistinguishably. We can even add higher order (`Float128`) methods as well based on this.

Though, it is the standard to cast lower floating point numbers to higher and do the calculation in higher precision (https://github.com/JuliaMath/SpecialFunctions.jl/blob/36c547b4a270b6089b1baf7bec05707e9fb8c7f9/src/beta_inc.jl#L16). Especially for `Float16` as they are just really for storage and not really dedicated hardware for this (it will be faster to do the calculation in `Float32` than do it in `Float16`. 

For `struveK` and `struveH` we need a bit more care as they are prone to cancellation. I will add dedicated Float32 routines for this, however they should also cast the Float32 number to Float64 and do the calculation so those expansions are valid in larger regions. I'm open to whatever you think on this but these routines probably will benefit from doing the calculations all in `Float64` to a desired tolerance.

gwater commented on src/struveL.jl:
```diff
@@ -0,0 +1,70 @@
+struveL(v, x) = _L_integral(v, x)
+
+_struveL(v, x::AbstractFloat) = _L_integral(v, x)
+_struveL(v::Float32, x::Float32) = Float32(_struveL(Float64(v), Float64(x)))
```
see my comment on `_struveH`

gwater commented on src/struveH.jl:
```diff
@@ -0,0 +1,73 @@
+end
+
+struveK(v::Real, x::Real) = _struveK(float(v), float(x))
+struveK(v, x) = _K_integral(v, x)
```
I think this sort of dispatch could be done in `src/Struve.jl` by defining multiple methods for `Struve.K` etc. I guess this raises the question: Is `Struve.K` okay for users or do we need to export something like `struvek` (in analogy to `besselj` from SpecialFunctions)

heltonmc replied:
I'm generally supported of exported `struvek` and `struveh` in the `src/Struve.jl` just to keep the ecosystem consistent. There has been some discussion on this at SpecialFunctions.jl for other functions and I think the general consensus is lower case for all the function similar to `besselk`. 

What do you think? I find the naming in the files when looking through the source easier to reason about if the function is `struvek` instead of `k` but that is probably because I'm also looking at other sources at the same time with `besselk` :smile:

gwater replied:
I agree, let's export `struvek`, `struvem` etc so that we're consistent with SpecialFunctions.jl . Accordingly, we should remove the constants for `Struve.K`, `Struve.M` etc (and bump the version number)

heltonmc replied:
Sounds great! I've changed these function definitions, added significant more docs and links to NIST, and incremented version. I will sort through the constant definitions and make sure I didn't miss anything this afternoon!

heltonmc replied:
Looks like I need to look through the subnormal cases as they sometimes fail. I’ll have time in the afternoon to sort those out.

heltonmc committed:
update types and avoid overflow

heltonmc committed:
update type

heltonmc committed:
add F32 for struvel and struvem

heltonmc committed:
add F32 for struveh and struvek

heltonmc commented:
I think I've sorted out `Float32` implementations now for all the methods. For `struvel` and `struvem` they can just be done within the series in Float32 format without the worry for cancellation. `struvek` and `struveh` we need to worry about cancellation more. For those series I've promoted the type to `Float64` and will sum until the desired tolerance is met either for F32 or F64. This increases the range of where we can use the series substantially so the fall back to numerical integration is rarely reached in Float32. This will results in very large speedups until we can avoid that integration fallback completely.

I need to add more Float32 test cases specifically now and see if we want to use all lower case and export the functions. I can go ahead and make those changes to.

heltonmc committed:
add docs, change function defintions, inc. version

heltonmc committed:
add chebyshev coefficients

heltonmc committed:
use 3 branches for chebyshev approximation

heltonmc committed:
add overflow checks and logs

heltonmc committed:
add logs for struveh and struvek

gwater commented on src/small_arg.jl:
```diff
@@ -1,11 +1,81 @@
 
-_inner_small(n, ν, z) = (0.5z)^(2n) / (gamma(n + 1.5) * gamma(n + ν + 1.5))
+const GAMMA_THREE_HALF(::Type{Float32}) = 0.886226925452758f0
+const GAMMA_THREE_HALF(::Type{Float64}) = 0.886226925452758
```
afaik functions are always constant in Julia

heltonmc committed:
fallback to integral.........

heltonmc committed:
update cutoff

heltonmc commented:
Ok... after a bit of work I think this is ready and all tests are passing on master. I'll try to summarize though I've added some docs to help out.
`struvel` and `struvem` are in some regards easy. The power series can always be used without cancellation to any desired precision so we can always just fallback to using that when the asymptotic expansions aren't valid. This is much faster (and more accurate) than using the numerical integral.
```julia
julia> @benchmark struvel(v, x) setup=(v=rand()*10; x=rand()*50)
BenchmarkTools.Trial: 10000 samples with 909 evaluations.
 Range (min … max):  114.042 ns … 507.306 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     205.208 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   216.175 ns ±  64.766 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

           ▁▂▂▂▄▅█▇▅▄▁▁▁ ▃▁▂                                     
  ▂▃▄▄▅▆▆▇▇█████████████████▇▃▂▂▂▂▁▁▁▁▁▁▂▁▁▂▁▁▁▁▂▁▁▁▁▁▁▂▂▃▄▅▄▃▃ ▄
  114 ns           Histogram: frequency by time          452 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark Struve._L_integral(v, x) setup=(v=rand()*10; x=rand()*50)
BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  1.548 μs … 263.102 μs  ┊ GC (min … max): 0.00% … 97.07%
 Time  (median):     7.522 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   8.258 μs ±   7.454 μs  ┊ GC (mean ± σ):  0.93% ±  1.68%

       ▁    ▁  █                                               
  ▄▂▂▁▄█▄▂▃██▂▃█▇▃▂▂▂▂▂▂▂▂▂▁▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▂▂▂▂▂▂▂▂▂▂▂ ▂
  1.55 μs         Histogram: frequency by time        35.4 μs <

 Memory estimate: 208 bytes, allocs estimate: 9.
```
In fact, the `struvel` time is largely dependent on the call to `besseli` for the large arg expansion. We could switch over to `Bessels.jl` which would decrease that time by about 5x.


On the other hand `struveh` and `struvek` are just tough. I will need to probably spend some more time on that if we want to fully remove the call to `_H_integral` which I unfortunately wasn't completely able to do for large orders (`nu>30`). Though for orders less than that the integral form is never called. Originally, I had thought that forward recurrence was stable for `struvek` and that if we could generate lower orders we could then generate all the higher orders. Now I spent some time using a Chebyshev approximation for low oders (nu < 2) and x < 39. For x>39 we can compute with the asymptotic expansion. It turns out that forward recurrence is only stable really up to orders like nu=30 so this approach worked well for that so I've kept it.

One thing I didn't try thoroughly is using backward recurrence but we will need that large order expansion for that so might work on that in a separate PR. The issue with using the power series for large orders is that it suffers from cancellation when nu ~ x so you need nu >> x for it to work. But that requires you to use recurrence over a lot of orders which is really slow. And honestly, we may come to a point where backward recurrence is also not stable so that might be mute anyway.

Alternatively, I also tried fixed point gauss integration but found you needed a lot of points (>150) to get good accuracy and because we wouldn't have any bound on x or nu because recurrence is unstable this method is probably not a good one. 
Finally there is the expansion in terms of bessel functions (http://dlmf.nist.gov/11.4.E20](https://dlmf.nist.gov/11.4.E20)) that might be worth another consideration. I briefly tried that but found it also suffered from some cancellation but might be worth it to re investigate.

gwater commented:
Looks great! Please add your name the [copyright notice](https://github.com/gwater/Struve.jl/blob/master/LICENSE.md) and then we're good to go.

heltonmc committed:
update license

heltonmc commented:
Great! Ok I updated it also so GitHub can recognize the MIT license. I'll look into expansions for large order and expansions of Bessel functions next. This should close #9, #2. 

Edit: I also updated the readme.  It might need a bit more but feel free to add or edit that more!

heltonmc committed:
update readme

heltonmc committed:
fix typo
